tensorstore/kvstore/s3/s3_request_builder.cc

```diff
@@ -0,0 +1,317 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <string>
+#include <string_view>
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/cord.h"
+#include "absl/status/status.h"
+#include "absl/time/time.h"
+#include "tensorstore/kvstore/s3/s3_request_builder.h"
+#include "tensorstore/kvstore/s3/validate.h"
+#include "tensorstore/internal/ascii_utils.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/util/result.h"
+#include "tensorstore/util/str_cat.h"
+
+using ::tensorstore::internal::AsciiSet;
+using ::tensorstore::internal::ParseGenericUri;
+using ::tensorstore::internal::ParsedGenericUri;
+using ::tensorstore::internal_storage_s3::IsValidBucketName;
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+
+namespace {
+
+// See description of function UriEncode at this URL
+// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+constexpr AsciiSet kUriUnreservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "-._~"};
+
+// NOTE: Only adds "/" to kUriUnreservedChars
+constexpr AsciiSet kUriObjectKeyReservedChars{
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789"
+    "/-._~"};
+}
+
+
+std::string UriEncode(std::string_view src) {
+    std::string dest;
+    PercentEncodeReserved(src, dest, kUriUnreservedChars);
+    return dest;
+}
+
+std::string UriObjectKeyEncode(std::string_view src) {
+    std::string dest;
+    PercentEncodeReserved(src, dest, kUriObjectKeyReservedChars);
+    return dest;
+}
+
+S3RequestBuilder::S3RequestBuilder(std::string_view method,
+                                   std::string endpoint_url)
+    : query_parameter_separator_("?") {
+  request_.method_ = method;
+  request_.url_ = std::move(endpoint_url);
+}
+
+HttpRequest S3RequestBuilder::BuildRequest(
+    std::string_view aws_access_key,
+    std::string_view aws_secret_access_key,
+    std::string_view aws_region,
+    std::string_view payload_hash,
+    const absl::Time & time) {
+
+
+  auto headers = request_.headers();
+  std::sort(std::begin(headers), std::end(headers));
+  std::sort(std::begin(encoded_queries_), std::end(encoded_queries_));
+  auto canonical_request = CanonicalRequest(request_.url(), request_.method(),
+                                            payload_hash, headers,
+                                            encoded_queries_);
+  auto signing_string = SigningString(canonical_request, aws_region, time);
+  auto signature = Signature(aws_secret_access_key, aws_region, signing_string, time);
+  auto auth_header = AuthorizationHeader(aws_access_key, aws_region, signature, headers, time);
+  request_.headers_.emplace_back(std::move(auth_header));
+  return std::move(request_);
+}
+
+S3RequestBuilder& S3RequestBuilder::AddUserAgentPrefix(std::string_view prefix) {
```

Comment:
For these methods you can delegate to request_.AddUserAgentPrefix, etc.

Reply:
Agreed, but I think this is going to be depend on whether `S3RequestBuilder` inherits/encapsulates from `HttpRequest`.

Reply:
Encapsulated `HttpRequestBuilder` in  `S3RequestBuilder`